api/proto/info/info.proto

```diff
@@ -840,6 +840,14 @@ message ZInfoStorage {
   bool storageLocation = 4; // Storage location for app disks, images etc.
 }
 
+// Per snapshot information
+message ZInfoSnapshot {
+  // identifier for the snapshot, required
+  string id = 1;
+  // an app instance configuration, encoded as bytes, optional
```

Comment:
Is this the content of `AppInstanceConfig` received at the time when the snapshot was made?

Reply:
Yep. In the form that the controller can use to store it as a value in the current_config/next_config fields.

Reply:
Why would we want to send back all of that config? It can be very large since it includes all of the ACLs etc.
If you want to track versions it would be better to include a sha or a timestamp in the App Insytance config message and echo that much shorter string here.

Reply:
That's what EVE stores locally, and that's what it will expect to receive from the Controller during the rollback process.
Not to add versioning of the tables or this field on the Controller side, we have decided just to send the latest version of pre-snapshotted config from the device. 
@xyuria-zededa fix me if I'm wrong.

Reply:
Added the explanations in the field description.

Reply:
Controller knows the config it needs to save without any help from the device. So I still don't understand why we would ever need this.

If we do need it we need to define its format (is it protobuf? json rendering of the protobuf? an internal EVE structure like AppInstanceConfig?) Note that neither of those have been designed to capture all of the app instance configuration because there are UUID references to other objects.

Further the size of info messages are limited to about 64kbyte and the size of the ACLs and the cloud init userdata in the app instance config are limited to 2Mbyte, so I don't see how you can serialize all of that and send it back.
But the fundamental issue is that I don't see how the controller would ever make use of this. Is the high-level controller operation as it relates to this information written down somewhere?

Reply:
Suppose the device triggers the snapshotting automatically. In this case, only the device knows what the latest config it has received was. So, we have to inform the controller about it.
In the case the snapshotting was triggered by the controller, we do not even use this field. So, we add it for future use on this iteration and do not handle it.

The format is similar to what the controller stores as the `next_config` on its side.
It was mentioned in the doc here: https://docs.google.com/document/d/1B-bB5HJ4c9aHX-wgcXjwuRv5El6vv4C-hLYZWUzkQxI/edit?disco=AAAAqBM4Qps

Reply:
> Suppose the device triggers the snapshotting automatically. In this case, only the device knows what the latest config it has received was. So, we have to inform the controller about it. In the case the snapshotting was triggered by the controller, we do not even use this field. So, we add it for future use on this iteration and do not handle it.

It is useful to have a way to pass a reference to the configuration which was used. But passing the whole content seems like a huge waste of bandwidth and not very helpful for the controller. Will it do ascii/utf-8/byte comparison of the whole content? What if the json->protobuf->json conversion at the two ends do not preserve the bytes (we know golang de-serialization into maps puts things in random order in the map!)

What problem do you see with passing a reference?

Reply:
> What problem do you see with passing a reference?

When we discussed it with @xyuria-zededa I had the impression that such references do not exist at the moment. And suppose we want to implement them (as I understand, it's equal to the implementation of the tables versioning). In that case, it will take much more work than we originally expected to put into a simple snapshotting solution. It's described here in the doc: https://docs.google.com/document/d/1B-bB5HJ4c9aHX-wgcXjwuRv5El6vv4C-hLYZWUzkQxI/edit?disco=AAAAo6ow5x4

Reply:
If we finally decide to pass the whole content, I can look for a similar logic in our code (if any exists). Or find anther production-ready solution.
But again, this field may be used only in the future, and I would prefer to avoid the lack of details here to block the first implementation.

Reply:
I will update it soon.